File: 백준/Platinum/13544. 수열과 쿼리 3/수열과 쿼리 3.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.StringTokenizer;
import java.util.stream.Stream;

public class Main {
	static int N;
	static int[] numbers;
	static ArrayList<Integer>[] tree;

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		StringBuilder sb = new StringBuilder();

		N = Integer.parseInt(br.readLine());
		numbers = Stream.of(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
		tree = new ArrayList[N * 2];

		init();

		int ans = 0;
		int M = Integer.parseInt(br.readLine());
		while (M-- > 0) {
			st = new StringTokenizer(br.readLine());
			int a = Integer.parseInt(st.nextToken()) ^ ans;
			int b = Integer.parseInt(st.nextToken()) ^ ans;
			int k = Integer.parseInt(st.nextToken()) ^ ans;
			ans = query(a + N - 1, b + N, k);
			sb.append(ans).append("\n");
		}
		System.out.println(sb);
	}

	static void init() {
		for (int i = 0; i < tree.length; i++)
			tree[i] = new ArrayList<>();
		for (int i = N; i < N + numbers.length; i++) {
			int n = i;
			do {
				tree[n].add(numbers[i - N]);
			} while ((n /= 2) != 0);
		}
		for (int i = 0; i < N; i++)
			Collections.sort(tree[i]);
	}

	static int query(int l, int r, int key) {
		int res = 0;
		for (; l < r; l >>= 1, r >>= 1) {
			if ((l & 1) == 1) {
				if (tree[l].get(0) > key)
					res += tree[l].size();
				else if (tree[l].get(tree[l].size() - 1) <= key)
					;
				else
					res += tree[l].size() - upperBound(tree[l], key);
				l++;
			}

			if ((r & 1) == 1) {
				--r;
				if (tree[r].get(0) > key)
					res += tree[r].size();
				else if (tree[r].get(tree[r].size() - 1) <= key)
					;
				else
					res += tree[r].size() - upperBound(tree[r], key);
			}
		}
		return res;
	}

	public static int upperBound(ArrayList<Integer> arr, int key) {
		int mid;
		int front = 0;
		int rear = arr.size();
		while (front < rear) {
			mid = (front + rear) / 2;
			if (arr.get(mid) <= key)
				front = mid + 1;
			else
				rear = mid;
		}
		return rear;
	}
}

File: 백준/Platinum/13544. 수열과 쿼리 3/수열과 쿼리 3.cc
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int N;
vector<int> numbers;
vector<vector<int>> tree;

void init() {
    tree.resize(N * 2);
    for (int i = 0; i < tree.size(); i++)
        tree[i] = vector<int>();
    for (int i = N; i < N + numbers.size(); i++) {
        int n = i;
        do {
            tree[n].push_back(numbers[i - N]);
        } while ((n /= 2) != 0);
    }
    for (int i = 0; i < N; i++)
        sort(tree[i].begin(), tree[i].end());
}

int query(int l, int r, int key) {
    int res = 0;
    for (; l < r; l >>= 1, r >>= 1) {
        if (l & 1) {
            if (tree[l][0] > key)
                res += tree[l].size();
            else if (tree[l][tree[l].size() - 1] <= key)
                ;
            else
                res += tree[l].size() - (upper_bound(tree[l].begin(),tree[l].end(), key)-tree[l].begin());
            l++;
        }

        if (r & 1) {
            --r;
            if (tree[r][0] > key)
                res += tree[r].size();
            else if (tree[r][tree[r].size() - 1] <= key)
                ;
            else
                res += tree[r].size() - (upper_bound(tree[r].begin(), tree[r].end(), key)-tree[r].begin());
        }
    }
    return res;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);
    cin >> N;

    numbers.resize(N);
    for (int i = 0; i < N; i++)
        cin >> numbers[i];

    init();

    int M;
    cin >> M;
    int last_ans = 0;
    while (M-- > 0) {
        int a, b, k;
        cin >> a >> b >> k;
        a = a ^ last_ans;
        b = b ^ last_ans;
        k = k ^ last_ans;
        last_ans = query(a + N - 1, b + N, k);
        cout << last_ans << "\n";
    }

    return 0;
}